File: ios/Classes/GetAppsPlugin.swift
import Flutter
import UIKit

public class GetAppsPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "mm_terekhin/get_apps_channel", binaryMessenger: registrar.messenger())
        let instance = GetAppsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any]
        
        switch call.method {
        case "getInstalledMessengers":
            getInstalledMessengers(result: result)
        case "openMessenger":
            openMessenger(args: arguments)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func getInstalledMessengers(result: @escaping FlutterResult) {
        var installedMessengers: [[String: Any]] = []
        
        messengers.forEach { messenger in
            let isInstalled = isAppInstalled(urlScheme: messenger.url)
            
            if (isInstalled) {
                let messenger = ["app_name": messenger.name, "type": messenger.type]
                
                installedMessengers.append(messenger)
            }
        }
        
        result(installedMessengers)
    }
    
    private func isAppInstalled(urlScheme: String) -> Bool {
        if let url = URL(string: urlScheme) {
            return UIApplication.shared.canOpenURL(url)
        }
        return false
    }
    
    private func openMessenger(args: [String: Any]?) {
        guard let type = args?["type"] as? String, let arg = args?["args"] as? String else {
            return }
        
        switch type {
        case "whatsApp":
            open(url: "whatsapp://send?phone=\(arg)")
        case "telegram":
            open(url: "tg://resolve?domain=\(arg)")
        default:
            break
        }
    }
    
    private func open(url: String) {
        if let url = URL(string: url), UIApplication.shared.canOpenURL(url) {
            UIApplication.shared.open(url)
        }
    }
}

enum MessengerType: String {
    case whatsApp
    case telegram
}

private class App {
    let name: String;
    let type: String;
    let url: String;
    
    init(name: String, type: String, url: String) {
        self.name = name
        self.type = type
        self.url = url
    }
}


private let messengers: [App] = [
    App(name: "whatsApp", type: MessengerType.whatsApp.rawValue, url: "whatsapp://"),
    App(name: "telegram", type: MessengerType.telegram.rawValue, url: "tg://"),
]

File: android/src/main/kotlin/com/mm_terekhin/get_apps/GetAppsPlugin.kt
package com.mm_terekhin.get_apps

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.*
import kotlin.collections.*
import android.net.Uri

/** GetAppsPlugin */
class GetAppsPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel =
            MethodChannel(flutterPluginBinding.binaryMessenger, "mm_terekhin/get_apps_channel")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val arguments = call.arguments as? Map<String, Any>

        when (call.method) {
            "getInstalledMessengers" -> getInstalledMessengers(result)
            "openMessenger" -> openMessenger(arguments)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun getInstalledMessengers(result: Result) {
        val installedMessengers = arrayListOf<Map<String, String>>()

        for (messenger in messengers) {
            val isInstalled = isAppInstalled(messenger.packageName)

            if (isInstalled) {
                val app = mapOf("app_name" to messenger.name, "type" to messenger.type)
                installedMessengers.add(app)
            }
        }

        result.success(installedMessengers)

    }

    private fun openMessenger(args: Map<String, Any>?) {
        val type = args?.get("type") as? String
        val arg = args?.get("args") as? String

        if (type != null && arg != null) {
            when (type) {
                "whatsApp" -> open("https://api.whatsapp.com/send?phone=$arg")
                "telegram" -> open("tg://resolve?domain=$arg")
                else -> {}
            }
        }
    }

    private fun open(url: String) {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context.startActivity(intent)
    }

    private fun isAppInstalled(packageName: String): Boolean {
        val packageManager = context.packageManager

        try {
            packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES)
            return true
        } catch (e: PackageManager.NameNotFoundException) {
            return false
        }
    }

}

class App {
    val name: String
    val type: String
    val packageName: String

    constructor(_name: String, _type: String, _packageName: String) {
        name = _name
        type = _type
        packageName = _packageName
    }
}

enum class MessengerType(val type: String) {
    WHATSAPP("whatsApp"),
    TELEGRAM("telegram")
}

private val messengers = arrayOf(
    App("whatsApp", MessengerType.WHATSAPP.type, "com.whatsapp"),
    App("telegram", MessengerType.TELEGRAM.type, "org.telegram.messenger")
)